General refresh of driver station best practices (#3105)
wpilibsuite/frc-docs: source/docs/software/driverstation/driver-station-best-practices.rst

3. Consider used laptops rather than new. The FRC\ |reg| Driver Station and dashboard software uses very few system resources, so you don't need to buy a new laptop -- instead, buy a cheap 4-5 year old used one. You might even get one donated by a used computer store in your area.
Dedicate a laptop to be used solely for driving. Many teams do. A dedicated machine allows you manage the configuration for one goal – being ready to compete at the field. Dedicated means no other software except the FRC-provided Driver Station software and associated Dashboard installed or running.

> .. note:: Before buying a used laptop ensure it is compatible with [Windows 11](https://support.microsoft.com/en-us/windows/windows-11-system-requirements-86c11283-ea52-4782-9efd-7674389a7ba3). For example, only Intel 8th generation core processors and later are compatible.

Reviewer: I feel like even with the addition of planning for the future with WIndows 11 down below, this provides more details, and it's only really a problem with buying used laptops. This is certainly a case where it's worth spending a little more for Windows 11 compatibility.

#.  Ports

    a. A built-in Ethernet port is highly preferred. Ensure that it's a full-sized port. The hinged Ethernet ports don't hold up to repeated use.

>     #. If the Ethernet port on your laptop is dodgy, either replace the laptop (recommended) or buy a USB Ethernet dongle from a reputable brand. Many teams find that USB Ethernet is less reliable than built-in Ethernet, primarily due to cheap hardware and bad drivers. The dongles given to rookies in the KOP have a reputation for working well.

Reviewer: I'd list the PN here. CD said: TP-Link UE330

    #. 2 USB ports minimum.  These can be USB-A or USB-C ports, depending on the peripherals you're using.

#.  A keyboard. It’s hard to quickly do troubleshooting on touch-only computers at the field.
#.  A solid-state disk (SSD), 256 GB or larger. If the laptop has a rotating disk, spend $50 and replace it with a SSD.

> #.  Supports Windows 11.  While Windows 10 is still supported for driver station use, acquiring a machine that can't run Windows 11 isn't forward thinking.

Reviewer: I think this is not strongly worded enough. The driver station for 2027+ hasn't been announced, but it's not hard to believe that it wouldn't support an EOL OS. Additionally, since Windows 10 won't receive security updates without a subscription, that should probably be mentioned.

#.  AMD or Intel CPU. The Driver Station software does not work on machines with ARM CPUs.

> #.  2.4 and 5 GHz Wi-Fi radio, with 6GHz optional.  6GHz Wi-Fi can be used in some of the practice driving configurations.   See `recommended practice configuration <https://frc-radio.vivid-hosting.net/overview/practicing-at-home>`__ for details.

Reviewer: Well, the recommended practice configuration doesn't need 2.4 ghz or 5ghz wifi either :-)

The [FIRST laptop recommendations](https://www.firstinspires.org/hubfs/web/program/all/computer-requirements.pdf?hsLang=en) list wifi 6E as "preferred". I think I'd go with something along those lines. If I was buying a new $1000+ laptop, I'd want it. I wouldn't let it stop me from buying a $300 used laptop though.

Maybe a pro/con list to help teams. Pro, not tethered to access point, less wear/tear on connectors. Con, not using ethernet like real field, more possibility for dodgy wifi drivers to interfere.

> #.  2.4 and 5 GHz Wi-Fi radio, with 6GHz optional.  The `recommended practice configuration <https://frc-radio.vivid-hosting.net/overview/practicing-at-home#vh-109-as-an-access-point-recommended>`__ does not require a 6GHz radio in the driver station laptop.

Reviewer: Well, 2.4 and 5ghz wifi aren't needed either :-)

In my case, if I was buying a brand new $1000+ laptop, having the flexibility to use it wirelessly would be worth it to me. If I was buying a $300 used laptop, I wouldn't bother. I might consider a pro/con (pro more flexible, con less like field setup) to help teams make their own decision.

The FIRST laptop recommendations list Wifi 6E as "preferred". https://www.firstinspires.org/hubfs/web/program/all/computer-requirements.pdf?hsLang=en

## Building your driver station

> It's best if you combine your driver station laptop and all needed peripherals into a `single unit <https://www.google.com/search?udm=2&q=frc+driver+station>`__.

Reviewer: I don't really like linking to a google search that can change over time. Maybe a CD thread like https://www.chiefdelphi.com/t/looking-for-driver-station-ideas/463986 or https://www.chiefdelphi.com/t/show-me-your-operator-station-driver-station/435030 although neither have as many examples as I'd like.

#. Bring a trusted USB and Ethernet cable for use connecting to the roboRIO.

## At The Competition

1. Turn off Windows firewall using :ref:`these instructions <docs/networking/networking-introduction/windows-firewall-configuration:Disabling Windows Firewall>`.
#. Turn off the Wi-Fi adapter, either using the dedicated hardware Wi-Fi switch or by using the Wi-Fi control in Quick Settings.
#. Charge the driver station when it’s in the pit.
#. Ensure everyone on the drive team knows the password. You'd be surprised at how often drivers arrive at the field without knowing the password for the laptop.
#. LabView users: Ensure your LabView code is deployed permanently and set to "run as startup", using the instructions in the LabView Tutorial. If you must deploy code every time you turn the robot on, you’re doing it wrong.
#. Limit web browsing to FRC related web sites. This minimizes the chance of getting malware during the competition.

> #. Don't plan on using internet access at the event to make software updates. There likely won't be any in the venue, and hotel Wi-Fi varies widely in quality. If you do need updates, contact a Control System Advisor in the pit.

Reviewer: Maybe this should move to before competition. If you're reading this at competition, it's too late.

## Before Each Match

1. Make sure the laptop is on and logged in prior to the end of the match before yours.
#. Close programs that aren’t needed during the match.

> #. Bring your laptop charger to the field. Power is provided in each player station.

Reviewer: Maybe add another sentence about why. Most people think about whether they'll run out of battery, but there's other advantages. For example Windows is getting more agressive about power saving on battery by doing things like network packet coalescing and lowering priorities for non-visible applications for power saving which has negative consequences in our near-realtime system.

Consider used laptops rather than new. The FRC\ |reg| Driver Station and dashboard software uses very few system resources, so you don't need to buy a new laptop -- instead, buy a cheap 4-5 year old used one.  There are thousands for sale every day on eBay. You might even get one donated by a used computer store.

### Laptop recommended features

1.  8GB of RAM or greater
#.  A display size of 13" or greater, with 1080p minimum resolution
#.  Ports

    a. A built-in Ethernet port is highly preferred. Ensure that it's a full-sized port. The hinged Ethernet ports don't hold up to repeated use.
    #. If the Ethernet port on your laptop is dodgy, either replace the laptop (recommended) or buy a USB Ethernet dongle from a reputable brand (e.g., TP-Link UE330). Many teams find that USB Ethernet is less reliable than built-in Ethernet, primarily due to cheap hardware and bad drivers. The dongles given to rookies in the KOP have a reputation for working well.
    #. 2 USB ports minimum.  These can be USB-A or USB-C ports, depending on the peripherals you're using.

#.  A keyboard. It’s hard to quickly do troubleshooting on touch-only computers at the field.

> #.  A solid-state disk (SSD), 256 GB or larger. If the laptop has a rotating disk, spend $50 and replace it with a SSD.

Reviewer 2: ```suggestion
#.  A solid-state disk (SSD), 256 GB or larger. If the laptop has a rotating disk, spend $50 and replace it with an SSD.
```

> #.  Supports Windows 11.  Given that it is at end of life, don't plan on driver station support for Windows 10 beyond the 2026 competitionyear.

Reviewer: fix typo

```suggestion
#.  Supports Windows 11.  Given that it is at end of life, don't plan on driver station support for Windows 10 beyond the 2026 competition year.
```

1. Install all Windows updates, then test the driver station with your robot. Keep running Windows Update until no updates show as pending.
#. Pause updates. Navigate to Start -> Settings -> Windows Update, then set Pause updates to cover the time of the competition.
#. Change "Active Hours" for Windows Updates to prevent updates from installing during competition hours. This helps if you forget to pause updates.  Navigate to Start -> Settings -> Windows Update -> Advanced options, then select Active Hours. Be sure to take timezone differences at your competition location into account.
#. Bring a trusted USB and Ethernet cable for use connecting to the roboRIO.

## At The Competition

1. Turn off Windows firewall using :ref:`these instructions <docs/networking/networking-introduction/windows-firewall-configuration:Disabling Windows Firewall>`.
#. Turn off the Wi-Fi adapter, either using the dedicated hardware Wi-Fi switch or by using the Wi-Fi control in Quick Settings.
#. Charge the driver station when it’s in the pit.
#. Ensure everyone on the drive team knows the password. You'd be surprised at how often drivers arrive at the field without knowing the password for the laptop.

> #. LabView users: Ensure your LabView code is deployed permanently and set to "run as startup", using the instructions in the LabView Tutorial. If you must deploy code every time you turn the robot on, you’re doing it wrong.

Reviewer 2: ```suggestion
#. LabVIEW users: Ensure your LabVIEW code is deployed permanently and set to "run as startup", using the instructions in the LabVIEW Tutorial. If you must deploy code every time you turn the robot on, you’re doing it wrong.
```
